W3ChampionsStatisticService/Achievements/AchievementRepositoryHandler.cs

```diff
@@ -0,0 +1,272 @@
+            var achievementsToAdd = UpdateCurrentPlayerAchievementList(playerAchievements.PlayerAchievementList);
+            for(int i = 0; i < achievementsToAdd.Count; i++){
+                achievementsToAdd[i] = await UpdateCurrentPlayerAchievement(achievementsToAdd[i], playerProfile, null, null, true);
+                playerAchievements.PlayerAchievementList.Add(achievementsToAdd[i]);
+            }
+            return playerAchievements;
+        } 
+
+        private async Task<Matchup> GetMatchupFromMatch(Match match){
+            var id = ObjectId.Parse(match.id);
+            var matchupDetail = await _matchRepository.LoadDetails(id);
+            return matchupDetail.Match;
+        }
+
+        public async Task<PlayerAchievements> GetPlayerAchievementsFromUI(string battleTag){
+            var needsSave = false;
+            var playerAchievements = await _achievementRepository.GetPlayerAchievements(battleTag);
+            if (playerAchievements == null){
+                var playerProfile = await _playerRepository.LoadPlayerProfile(battleTag);
+                if (playerProfile != null){
+                    playerAchievements = await CreateNewPlayerAchievements(playerProfile);
+                    needsSave = true;
+                } else {
+                    return null;
+                } 
+            } else {
+                if (playerAchievements.PlayerAchievementList.Count < ActiveAchievementIds.Length){
+                    var achievementsToAdd = UpdateCurrentPlayerAchievementList(playerAchievements.PlayerAchievementList);
+                    var playerProfile = await _playerRepository.LoadPlayerProfile(battleTag);
+                    playerAchievements = await AddAdditionalAchievements(playerAchievements, playerProfile);
+                    needsSave = true;
+                }
+            }
+            if(needsSave){await _achievementRepository.UpsertPlayerAchievements(playerAchievements);}
+            return playerAchievements;
+        }
+
+        private List<int> ConvertSeasonsToSimpleList(List<Season> seasons){
+            var seasonList = new List<int>();
+            foreach (Season s in seasons){seasonList.Add(s.Id);}
+            seasonList.Reverse();
+            return seasonList;
+        }
+
+        private List<Achievement> UpdateCurrentPlayerAchievementList(List<Achievement> currentAchievementsList){
+            if (currentAchievementsList == null) {currentAchievementsList = new List<Achievement>();}
+                var currentListIds = new List<long>();
+                foreach(Achievement achievement in currentAchievementsList){
+                    currentListIds.Add(achievement.Id);
+                }
+                foreach(long activeAchievementsId in ActiveAchievementIds){
+                    if(!currentListIds.Contains(activeAchievementsId)){
+                        switch (activeAchievementsId) {
+                            case 0:
```

Comment:
if you have an abstract class/interface, those switch cases can go. This tends to explode heavily, would be 100 cases for 100 achievements.